src/Native/Unix/System.Native/pal_interfaceaddresses.c

```diff
@@ -195,15 +196,48 @@ int32_t SystemNative_EnumerateGatewayAddressesForInterface(uint32_t interfaceInd
         int isGateway = flags & RTF_GATEWAY;
         int gatewayPresent = hdr->rtm_addrs & RTA_GATEWAY;
 
-        if (isGateway && gatewayPresent)
+        if (isGateway && gatewayPresent && ((int)interfaceIndex == -1 || interfaceIndex == hdr->rtm_index))
         {
             IpAddressInfo iai;
+            struct sockaddr_storage* sock = (struct sockaddr_storage*)(hdr + 1);
             memset(&iai, 0, sizeof(IpAddressInfo));
-            iai.InterfaceIndex = interfaceIndex;
-            iai.NumAddressBytes = NUM_BYTES_IN_IPV4_ADDRESS;
-            struct sockaddr_in* sain = (struct sockaddr_in*)(hdr + 1);
-            sain = sain + 1; // Skip over the first sockaddr, the destination address. The second is the gateway.
-            memcpy_s(iai.AddressBytes, sizeof_member(IpAddressInfo, AddressBytes), &sain->sin_addr.s_addr, sizeof(sain->sin_addr.s_addr));
+            iai.InterfaceIndex = hdr->rtm_index;
+
+            if (sock->ss_family == AF_INET)
+            {
+                iai.NumAddressBytes = NUM_BYTES_IN_IPV4_ADDRESS;
+                struct sockaddr_in* sain = (struct sockaddr_in*)sock;
+                if (sain->sin_addr.s_addr != 0)
+                {
+                    // filter out normal routes.
```

Comment:
Can you help me understand this better?  Elaborating on the comment to explain what it means for this to be a normal route, how we know that from s_addr != 0, and why we're doing that filtering would be helpful.

Reply:
The API we use essentially dumps routing table. 
The route entry has prefix and nexthop. From the example I give on top the 11.0.0.0 would  turn into 0xb000 and 0.0.0.0 would turn to 0. So this (and memcmp bellow) filters out only "default routes".

Technically gateway could be any host willing to forward traffic not destined for it self. That may be all or some subnets. From testing .Net on Windows interpretation of gateway seems to be default only so the filter looks for ::0 or 0.0.0.0 prefixes.